### PR TITLE
[#7] [Integrate] As a user, I can see My Coins list on Home page

### LINKED ISSUE #7: [Integrate] As a user, I can see My Coins list on Home page
## Why

When the user launches the app, it will show the "My Coins" section on the Home page.

## Acceptance Criteria

- Mapping the data from API response to display on UI:
image
  - If the field `price_change_percentage_24h_in_currency`:
    - Positive number: Use green color with arrow up `#10DC78`
    - Negative number: Use red color with arrow down `#F15950`
- Use [Coil](image) to load the coin image

## Design

N/A

## Resources

N/A

### PR DESCRIPTION
**Pull request by luongvo**
Resolve https://github.com/nimblehq/jetpack-compose-crypto/issues/7

## What happened 👀

- Add HomeViewModel to request my coins data.
- Integrate VM to `HomeScreen` and  bind my coins data list.
- `Loading` and `error` showing will be integrated later 🏃 .

## Insight 📝

- `lifecycle-viewmodel-compose:2.5.1` requires to upgrade SDK to API `31`.
  - API 31 and modern libs require to upgrade to `Gradle 7` ✅ .
    - Android Gradle 7 plugin requires Java 11 to run https://github.com/nimblehq/jetpack-compose-crypto/runs/8184091690?check_suite_focus=true.
- `androidx.test` has issues about missing `android:exported` on API 31 https://issuetracker.google.com/issues/191397778, I ended up with having a workaround in manifest for temporary 😮‍💨 .
- Use [turbine](https://github.com/cashapp/turbine) to test Flow in ViewModel 🧑‍🍳 .

## Proof Of Work 📹

https://user-images.githubusercontent.com/16315358/188397748-31ff2b28-5958-4685-b04b-d662977fa3d6.mp4

### CODE REVIEW

File: app/src/test/java/co/nimblehq/compose/crypto/ui/screens/BaseViewModelTest.kt [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import kotlinx.coroutines.test.TestCoroutineScope
+import org.junit.Rule
+
+@ExperimentalCoroutinesApi
+abstract class BaseViewModelTest {
+
+    @get:Rule
+    private var coroutineRule = CoroutineTestRule()
+
+    protected fun runBlockingTest(block: suspend TestCoroutineScope.() -> Unit) =
```
Comment (luongvo) [inline app/src/test/java/co/nimblehq/compose/crypto/ui/screens/BaseViewModelTest.kt:15]: FYI: I will migrate these deprecated test utils in next integration PRs 🙏

File: data/build.gradle.kts [hunk 2]
```diff
@@ -10,8 +10,6 @@ android {
     defaultConfig {
         minSdk = Versions.ANDROID_MIN_SDK_VERSION
         targetSdk = Versions.ANDROID_TARGET_SDK_VERSION
-        versionCode = Versions.ANDROID_VERSION_CODE
-        versionName = Versions.ANDROID_VERSION_NAME
```
Comment (luongvo) [inline data/build.gradle.kts:14]: FYI: Gradle 7 does not have these config in Android Library.

File: .github/workflows/deploy_to_firebase_app_distribution.yml [hunk 3]
```diff
@@ -13,10 +13,11 @@ jobs:
     runs-on: ubuntu-latest
     timeout-minutes: 30
     steps:
-      - name: Set up JDK 1.8
-        uses: actions/setup-java@v1
+      - name: Set up JDK 11
+        uses: actions/setup-java@v3
         with:
-          java-version: 1.8
+          distribution: 'temurin'
```
Comment (hoangnguyen92dn) [inline .github/workflows/deploy_to_firebase_app_distribution.yml:19]: Can you please explain why do we need to add this configuration? 🤔

File: app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+                )
+            )
+                .catch { e ->
+                    _error.emit(e)
+                }
+                .collect { myCoins ->
+                    _myCoins.emit(myCoins.map { it.toUiModel() })
+                }
+            hideLoading()
+        }
```
Comment (hoangnguyen92dn) [inline app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModel.kt]: 🤔 The list of my coins but be requested on init? How about creating a public function and execute from the HomeScreen?

File: app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    val myCoins: StateFlow<List<CoinItemUiModel>>
+}
+
+@HiltViewModel
+class HomeViewModel @Inject constructor(
+    dispatchers: DispatchersProvider,
+    private val getMyCoinsUseCase: GetMyCoinsUseCase
+) : BaseViewModel(dispatchers), BaseInput, Output {
+
+    override val input = this
```
Comment (hoangnguyen92dn) [inline app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModel.kt:28]: How about
```suggestion
override val input: BaseInput
        get() = this
```

File: app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/CoinItem.kt [hunk 6]
```diff
@@ -97,22 +97,24 @@ fun CoinItem(
...
+            text = stringResource(
+                R.string.coin_currency,
+                coinItem.currentPrice.toFormattedString()
+            ),
             color = MaterialTheme.colors.textColor,
             style = Style.semiBold16()
         )
 
         PriceChange(
+            coinItem = coinItem,
```
Comment (hoangnguyen92dn) [inline app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/CoinItem.kt]: Should we avoid passing the model into this component? 🤔 I see that we only need `priceChangePercentage24hInCurrency` property.

File: .github/workflows/deploy_to_firebase_app_distribution.yml (hunk 3 above)
Comment (luongvo) [reply .github/workflows/deploy_to_firebase_app_distribution.yml:19]: @hoangnguyen92dn Gradle 7 requires Java 11 to run 🤓 https://github.com/nimblehq/jetpack-compose-crypto/runs/8184091690?check_suite_focus=true. Let me add this info to PR description.

File: app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModel.kt (hunk 4 above)
Comment (luongvo) [reply app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModel.kt]: @hoangnguyen92dn we don't have plan for something like reload or pull down to refresh. Loading data at init is fine for now @hoangnguyen92dn.

Please plan to show a `loading` or `error message` in a new ticket. 🙏

File: app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModel.kt (hunk 5 above)
Comment (luongvo) [reply app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModel.kt:28]: @hoangnguyen92dn my approach `override val input = this` means I will assign `this` to a new `val` at **initial time**. While yours means passing `this` to `val` at **calling time**. They're different, and
- I would like to assign it at the initial time because we don't do to extra operations there.
- `this` is a reference, the value for both cases are the same.

File: app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/CoinItem.kt (hunk 6 above)
Comment (luongvo) [reply app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/CoinItem.kt]: @hoangnguyen92dn changeable 👍

Comment (luongvo) [reply app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/CoinItem.kt]: @hoangnguyen92dn Fixed in https://github.com/nimblehq/jetpack-compose-crypto/pull/27/commits/95a4d72049218676751105b68f09127f25a718f7

File: app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+private const val MY_COINS_ITEM_PER_PAGE = 10
+
+interface Output : BaseOutput {
+    val myCoins: StateFlow<List<CoinItemUiModel>>
+}
+
+@HiltViewModel
+class HomeViewModel @Inject constructor(
+    dispatchers: DispatchersProvider,
+    private val getMyCoinsUseCase: GetMyCoinsUseCase
```
Comment (Wadeewee) [inline app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModel.kt]: I don't think we need `private val`, if we will call it only in `init` scope 🤔 
```suggestion
    getMyCoinsUseCase: GetMyCoinsUseCase
```

File: app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModel.kt [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+                    itemPerPage = MY_COINS_ITEM_PER_PAGE,
+                    page = 1
+                )
+            )
+                .catch { e ->
+                    _error.emit(e)
+                }
+                .collect { myCoins ->
+                    _myCoins.emit(myCoins.map { it.toUiModel() })
+                }
```
Comment (Wadeewee) [inline app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModel.kt:52]: Let's format like this:
```suggestion
            ).catch { e ->
                _error.emit(e)
            }.collect { myCoins ->
                _myCoins.emit(myCoins.map { it.toUiModel() })
            }

```

File: app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModel.kt [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    init {
+        execute {
+            showLoading()
+            getMyCoinsUseCase.execute(
+                GetMyCoinsUseCase.Input(
+                    currency = MY_COINS_CURRENCY,
+                    order = MY_COINS_ORDER,
+                    priceChangeInHour = MY_COINS_PRICE_CHANGE_IN_HOUR,
+                    itemPerPage = MY_COINS_ITEM_PER_PAGE,
+                    page = 1
```
Comment (Wadeewee) [inline app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModel.kt]: Should we create a `const val` for this ☝️  May be `MY_COINS_DEFAULT_PAGE`

File: app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModel.kt (hunk 4 above)
Comment (Wadeewee) [reply app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModel.kt]: I think It would be better if we separate to function to minimize the code in the `init` scope when calling for `getCoins` and `getTrending` 🤔

File: app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModel.kt (hunk 8 above)
Comment (luongvo) [reply app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModel.kt:52]: @Wadeewee tbh, I don't like that format. we usually start a chain call with a dot `.`, e.g.
```
coins
  .filter { ... }
  .sort { ... }
  .toArray()
```

File: app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/CoinItem.kt [hunk 10]
```diff
@@ -131,6 +133,15 @@ fun CoinItemPreview() {
 @Preview
 fun CoinItemPreviewDark() {
     ComposeTheme(darkTheme = true) {
-        CoinItem()
+        CoinItem(coinItemPreview)
     }
 }
+
+internal val coinItemPreview = CoinItemUiModel(
```
Comment (Wadeewee) [inline app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/CoinItem.kt:140]: Shall we move to `CoinItemUiModel` class. 🤔  As It was called in many places.

Comment (luongvo) [reply app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/CoinItem.kt:140]: @Wadeewee it's not production code, I'm still finding a way to manage it better. One of the ways is https://developer.android.com/jetpack/compose/tooling#preview-multipreview 🤔 

Let me try to refactor it in the next PR.

File: app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModel.kt (hunk 5 above)
Comment (hoangnguyen92dn) [reply app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModel.kt:28]: @luongvo Yup. But we actually don't use `setter` function, so that's why I suggest to `return this` on the `getter` instead.

File: app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModel.kt (hunk 4 above)
Comment (hoangnguyen92dn) [reply app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModel.kt]: @luongvo I agree with @Wadeewee, but you can update on the next integration. It works for me 🙏

**Review by hoangnguyen92dn — APPROVED** [review]
Rest LGTM. Some changes are optional and can apply on the next integration PR.

File: app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModel.kt (hunk 5 above)
Comment (luongvo) [reply app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModel.kt:28]: @Wadeewee will refactor it in the next integration 👌

File: app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModel.kt (hunk 4 above)
Comment (luongvo) [reply app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModel.kt]: @Wadeewee I will refactor it in the next integration 👌

File: app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModel.kt (hunk 8 above)
Comment (Wadeewee) [reply app/src/main/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModel.kt:52]: Got it and noted @luongvo , I will use this pattern then 🚀

File: app/src/test/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModelTest.kt [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    @After
+    fun tearDown() {
+        Dispatchers.resetMain()
+        testDispatcher.cleanupTestCoroutines()
+    }
+
+    @Test
+    fun `When getting my coin list successfully, it should return my coin list`() =
+        runBlockingTest {
+            testDispatcher.pauseDispatcher()
```
Comment (Wadeewee) [inline app/src/test/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModelTest.kt:41]: It seems `deprecated` here. Did you get like this 👇  on your side @luongvo 
Screen Shot 2565-09-06 at 19 04 36

File: app/src/test/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModelTest.kt [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+                testDispatcher.resumeDispatcher()
+                expectMostRecentItem() shouldBe expected
+                cancelAndIgnoreRemainingEvents()
+            }
+        }
+
+    @Test
+    fun `When getting my coin list failed, it should throw error`() =
+        runBlockingTest {
+            testDispatcher.pauseDispatcher()
```
Comment (Wadeewee) [inline app/src/test/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModelTest.kt:54]: Question @luongvo Why do we need to call `pauseDispatcher()` before `initViewModel`, It doesn't clear when `cleanupTestCoroutines()` in `tearDown()` 🤔

File: app/src/test/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModelTest.kt (hunk 11 above)
Comment (luongvo) [reply app/src/test/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModelTest.kt:41]: @Wadeewee As I described here https://github.com/nimblehq/jetpack-compose-crypto/pull/27#discussion_r962587131, I will resolve it in next PRs.

File: app/src/test/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModelTest.kt (hunk 12 above)
Comment (luongvo) [reply app/src/test/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModelTest.kt:54]: @Wadeewee to pause the vm dispatcher to mock failed response for GetMyCoinsUsecase. Because we're executing it at init block :nerd_face:

File: app/src/test/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModelTest.kt [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.flowOf
+import kotlinx.coroutines.test.resetMain
+import kotlinx.coroutines.test.setMain
+import org.junit.*
+
+@ExperimentalCoroutinesApi
+class HomeViewModelTest : BaseViewModelTest() {
```
Comment (Wadeewee) [inline app/src/test/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModelTest.kt:20]: I appreciate you for adding the `Unit Test.` 👏  but I think It would be better if we add it later after creating the setup or updating the base class (`BaseViewModelTest`). So we can get rid of adding dependencies of `Unit Test` in `build.gradle`. What do you think? 🙇‍♀️

Comment (Wadeewee) [reply app/src/test/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModelTest.kt:20]: OK, Let's see on next integration  https://github.com/nimblehq/jetpack-compose-crypto/pull/27#discussion_r963652707

File: app/src/test/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModelTest.kt (hunk 12 above)
Comment (Wadeewee) [reply app/src/test/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModelTest.kt:54]: Oh, I see, If that, Would it be better if we execute from `HomeScreen` regarding this [comment](https://github.com/nimblehq/jetpack-compose-crypto/pull/27#discussion_r9632202041) 🤔

File: app/src/main/java/co/nimblehq/compose/crypto/extension/BigDecimalExt.kt [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+    maxDigit: Int = 2
+): String {
+    return try {
+        DecimalFormat(format).apply {
+            minimumIntegerDigits = 1
+            minimumFractionDigits = minDigit
+            maximumFractionDigits = maxDigit
+        }.format(this)
+    } catch (e: IllegalArgumentException) {
+        ""
```
Comment (Wadeewee) [inline app/src/main/java/co/nimblehq/compose/crypto/extension/BigDecimalExt.kt:18]: Should we return `"0"` 🤔

File: app/src/main/java/co/nimblehq/compose/crypto/extension/DoubleExt.kt [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+    maxDigit: Int = 2
+): String {
+    return try {
+        DecimalFormat(format).apply {
+            minimumIntegerDigits = 1
+            minimumFractionDigits = minDigit
+            maximumFractionDigits = maxDigit
+        }.format(this)
+    } catch (e: IllegalArgumentException) {
+        ""
```
Comment (Wadeewee) [inline app/src/main/java/co/nimblehq/compose/crypto/extension/DoubleExt.kt:17]: Same here should we return `"0.0"` 🤔

File: domain/src/main/java/co/nimblehq/compose/crypto/domain/model/CoinItem.kt [hunk 16]
```diff
@@ -1,18 +1,21 @@
...
     val marketCapRank: Int,
-    val totalVolume: Long,
-    val high24h: Long,
-    val low24h: Long,
-    val priceChange24h: Double,
+    val totalVolume: BigDecimal,
+    val high24h: BigDecimal,
+    val low24h: BigDecimal,
+    val priceChange24h: BigDecimal,
     val priceChangePercentage24h: Double,
```
Comment (Wadeewee) [inline domain/src/main/java/co/nimblehq/compose/crypto/domain/model/CoinItem.kt:17]: Regarding [this](https://github.com/nimblehq/jetpack-compose-crypto/pull/26#discussion_r963222963) Shall we change the type from `Double` to `BigDecimal,` same for other fields? I'm good. If you want to update on the next `PR` 🚀

File: app/src/test/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModelTest.kt (hunk 12 above)
Comment (luongvo) [reply app/src/test/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModelTest.kt:54]: @Wadeewee it's not really better I think.
- As because we don't have a human interaction like Retry or pull down to refresh the my coins data so we don't need to create an input to load it from HomeScreen.
- Because we load it at init block, we need to use pause dispatcher to mock the test response. I don't think this is a bad thing as Test dispatcher is built for this case of testing. I don't think changing the production code to just make the test passed is a great idea.

File: app/src/main/java/co/nimblehq/compose/crypto/extension/BigDecimalExt.kt (hunk 14 above)
Comment (luongvo) [reply app/src/main/java/co/nimblehq/compose/crypto/extension/BigDecimalExt.kt:18]: @Wadeewee this is the format func, I don't think 0 make sense here in case of error.

File: domain/src/main/java/co/nimblehq/compose/crypto/domain/model/CoinItem.kt (hunk 16 above)
Comment (luongvo) [reply domain/src/main/java/co/nimblehq/compose/crypto/domain/model/CoinItem.kt:17]: @Wadeewee I only recommend to use BigDecimal for pricing fields. This is percentage field so it's not the same thing, double is enough.

File: app/src/test/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModelTest.kt (hunk 13 above)
Comment (luongvo) [reply app/src/test/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModelTest.kt:20]: @Wadeewee I don't like having business logic without tests. The current VM and networking layers are good enough for this compose project. Hence we should have unit tests.

File: app/src/main/java/co/nimblehq/compose/crypto/extension/BigDecimalExt.kt (hunk 14 above)
Comment (Wadeewee) [reply app/src/main/java/co/nimblehq/compose/crypto/extension/BigDecimalExt.kt:18]: Hmm 🤔 But we will call in the `UI` screen. Is that mean when we have an error, we will show just `$` or `%`? e.g.   `CoinItem`

Comment (luongvo) [reply app/src/main/java/co/nimblehq/compose/crypto/extension/BigDecimalExt.kt:18]: @Wadeewee Thanks for raising, that's an edge case, we should take care of that. Let's do it later, we don't have much time to do everything in just 1 PR 🙏

File: app/src/test/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModelTest.kt (hunk 13 above)
Comment (Wadeewee) [reply app/src/test/java/co/nimblehq/compose/crypto/ui/screens/home/HomeViewModelTest.kt:20]: I totally agree with you, @luongvo, but It seems like the base test case needs to be updated. That's why I think we can skip the `Unit Test` for this PR because It has a lot of files that need to focus on e.g. `CoroutineTestRule` 🙏

**Review by Wadeewee — APPROVED** [review]
